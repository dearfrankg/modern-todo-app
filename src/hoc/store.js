import React, { Component } from "react";
import { connect } from "react-redux";

export default function(ComposedComponent, reducers) {
  class decoreWithStore extends Component {
    componentWillMount() {
      // eslint-disable-next-line
      // console.log('Will Mount HOC', reducers);
    }

    render() {
      return <ComposedComponent {...this.props} />;
    }
  }

  function mapStateToProps(state) {
    // eslint-disable-next-line
    if (reducers.length === 0) console.error("There' no reducer here, fix it!", reducers);

    const reducerMap = {};

    reducers.forEach(reducer => {
      reducerMap[reducer] = state[reducer];
    });

    return reducerMap;
  }

  return connect(mapStateToProps)(decoreWithStore);
}
