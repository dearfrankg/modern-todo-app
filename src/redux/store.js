import { createStore, combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import rootReducer from 'redux/reducers';
import { loadState, saveState } from 'redux/localStorage';

const persistedState = loadState();

const store = createStore(
  combineReducers({
    todos: rootReducer,
    router: routerReducer,
  }),
  persistedState,
);

store.subscribe(() => {
  saveState({
    todos: store.getState().todos,
  });
});

export default store;
