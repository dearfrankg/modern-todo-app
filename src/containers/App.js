import React from "react";
import { arrayOf, any, func } from "prop-types";
import { connect } from "react-redux";

import { Header, TodoList, Footer } from "containers";
import { Info } from "components";

import { completeAll as completeAllAction } from "redux/actions";
import { getAllCompleted } from "helpers";
import withStore from "hoc";

const propTypes = {
  todos: arrayOf(any).isRequired,
  completeAll: func.isRequired
};

export const AppContainer = ({ todos, completeAll }) => {
  const hasTodos = todos.length > 0;
  const isAllChecked = getAllCompleted(todos);

  return (
    <div>
      <section className="todoapp">
        <Header />

        {hasTodos && (
          <div>
            <section className="main">
              <input
                id="toggle-all"
                className="toggle-all"
                type="checkbox"
                onChange={completeAll}
                checked={isAllChecked}
              />
              <label htmlFor="toggle-all">Mark all as complete</label>

              <TodoList />
            </section>
            <Footer />
          </div>
        )}
      </section>

      <Info />
    </div>
  );
};

export const mapDispatchToProps = dispatch => ({
  completeAll: () => dispatch(completeAllAction())
});

AppContainer.propTypes = propTypes;

export default connect(
  null,
  mapDispatchToProps
)(withStore(AppContainer, ["todos"]));
