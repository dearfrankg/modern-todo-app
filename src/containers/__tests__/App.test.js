import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';

import store from 'redux/store';
import App, { AppContainer, mapDispatchToProps } from '../App';

describe('App Container:', () => {
  it('should render correctly with default props', () => {
    expect(
      shallow(
        <Provider store={store}>
          <App />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly with todos', () => {
    expect(
      shallow(
        <AppContainer todos={[{ id: 1, completed: false, text: 'text' }]} completeAll={() => {}} />,
      ),
    ).toMatchSnapshot();
  });

  describe('should call actions', () => {
    const dispatchSpy = jest.fn();
    const { completeAll } = mapDispatchToProps(dispatchSpy);

    it('should call completeAll', () => {
      completeAll();
      const spyLastCall = dispatchSpy.mock.calls[0][0];
      expect(spyLastCall.type).toEqual(dispatchSpy.mock.calls[0][0].type);
    });
  });
});
