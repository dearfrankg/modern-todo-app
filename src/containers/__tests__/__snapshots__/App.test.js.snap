// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Container: should render correctly with default props 1`] = `<Connect(Connect(decoreWithStore)) />`;

exports[`App Container: should render correctly with todos 1`] = `
<div>
  <section
    className="todoapp"
  >
    <Connect(HeaderContainer) />
    <div>
      <section
        className="main"
      >
        <input
          checked={false}
          className="toggle-all"
          id="toggle-all"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="toggle-all"
        >
          Mark all as complete
        </label>
        <Connect(decoreWithStore) />
      </section>
      <Connect(Connect(decoreWithStore)) />
    </div>
  </section>
  <Info />
</div>
`;
